File: microservices/auto_processing_datasets/src/lib/database/mongodb/MongoDatabase.py
from pymongo import MongoClient
from typing import Optional
from ..base import BaseAdapter
import sys
import os

# Add src to path for absolute imports
_current_dir = os.path.dirname(os.path.abspath(__file__))
_src_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(_current_dir))), 'src')
if _src_dir not in sys.path:
    sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(_current_dir))))


class MongoDatabase(BaseAdapter):
    """MongoDB adapter implementation."""
    
    def __init__(self, host: str = None, port: int = None, dbname: str = None, 
                 username: str = None, password: str = None):
        from src.configs.env import DB_HOST, DB_PORT, DB_NAME, DB_USER, DB_PASSWORD
        self.host = host or DB_HOST
        self.port = port or DB_PORT
        self.dbname = dbname or DB_NAME
        self.username = username or DB_USER
        self.password = password or DB_PASSWORD
        self.client: Optional[MongoClient] = None
        self.db = None
    
    def _build_connection_uri(self) -> str:
        """Build MongoDB connection URI."""
        if self.username and self.password:
            return f"mongodb://{self.username}:{self.password}@{self.host}:{self.port}/{self.dbname}?authSource=admin"
        else:
            return f"mongodb://{self.host}:{self.port}/{self.dbname}"
    
    def connect(self) -> None:
        """Connect to MongoDB."""
        try:
            # Close existing connection if any
            if self.client:
                try:
                    self.client.close()
                except:
                    pass
            
            uri = self._build_connection_uri()
            # Only pass authSource if credentials are provided
            # PyMongo doesn't accept None for authSource
            if self.username and self.password:
                self.client = MongoClient(uri, authSource='admin', serverSelectionTimeoutMS=5000)
            else:
                self.client = MongoClient(uri, serverSelectionTimeoutMS=5000)
            self.db = self.client[self.dbname]
            
            # Test the connection
            self.client.admin.command('ping')
            print(f"Connected to MongoDB: {self.host}:{self.port}/{self.dbname}")
        except Exception as error:
            print(f"MongoDB connection error: {error}")
            self.client = None
            self.db = None
            raise
    
    def is_connected(self) -> bool:
        """Check if database is connected and connection is alive."""
        try:
            if self.client is None or self.db is None:
                return False
            # Ping the server to check if connection is alive
            self.client.admin.command('ping')
            return True
        except Exception:
            return False
    
    def ensure_connected(self) -> None:
        """Ensure database is connected, reconnect if necessary."""
        if not self.is_connected():
            print("Database connection lost, reconnecting...")
            self.connect()
    
    def disconnect(self) -> None:
        """Disconnect from MongoDB."""
        if self.client:
            try:
                self.client.close()
            except:
                pass
            self.client = None
            self.db = None
            print("Disconnected from MongoDB")
    
    def find_one(self, collection: str, filter: dict) -> dict | None:
        """Find one document in a collection."""
        self.ensure_connected()
        return self.db[collection].find_one(filter)
    
    def update_one(self, collection: str, filter: dict, update: dict) -> None:
        """Update one document in a collection."""
        self.ensure_connected()
        self.db[collection].update_one(filter, {"$set": update})

File: backend/src/lib/database/mongodb/MongoDatabase.ts
import { MongoClient, Db, Collection, MongoClientOptions } from 'mongodb';
import { DatabaseAdapter } from '../base/BaseAdapter';
import { DatabaseConfig } from '../types';

export class MongoDatabase implements DatabaseAdapter {
  private client: MongoClient | null = null;
  private db: Db | null = null;
  private config: DatabaseConfig;

  constructor(config: DatabaseConfig) {
    this.config = config;
  }

  private buildConnectionUri(): string {
    const { host, port, dbname, auth } = this.config;

    if (auth?.username && auth?.password) {
      // Connection with authentication - use admin database for auth
      return `mongodb://${encodeURIComponent(auth.username)}:${encodeURIComponent(auth.password)}@${host}:${port}/${dbname}?authSource=admin`;
    } else {
      // Connection without authentication
      return `mongodb://${host}:${port}/${dbname}`;
    }
  }

  

  async connect(): Promise<void> {
    try {
      const uri = this.buildConnectionUri();
      const options: MongoClientOptions = {
        ...this.config.options,
        authSource: this.config.auth ? 'admin' : undefined,
      };

      this.client = new MongoClient(uri, options);
      await this.client.connect();
      this.db = this.client.db(this.config.dbname);
      console.log(`Connected to MongoDB: ${this.config.host}:${this.config.port}/${this.config.dbname}`);
    } catch (error) {
      console.error('MongoDB connection error:', error);
      throw error;
    }
  }

  async disconnect(): Promise<void> {
    if (this.client) {
      await this.client.close();
      this.client = null;
      this.db = null;
      console.log('Disconnected from MongoDB');
    }
  }


  async createCollection(name: string): Promise<void> {
    if (!this.db) {
      throw new Error('Database not connected');
    }
    try {
      const collections = await this.db.listCollections({ name }).toArray();
      if (collections.length === 0) {
        await this.db.createCollection(name);
        console.log(`Collection ${name} created`);
      }
    } catch (error) {
      console.error(`Error creating collection ${name}:`, error);
      throw error;
    }
  }

  async collectionExists(name: string): Promise<boolean> {
    if (!this.db) {
      throw new Error('Database not connected');
    }
    const collections = await this.db.listCollections({ name }).toArray();
    return collections.length > 0;
  }

  private getCollection(collection: string): Collection {
    if (!this.db) {
      throw new Error('Database not connected');
    }
    return this.db.collection(collection);
  }

  async insertOne(collection: string, document: any): Promise<any> {
    const result = await this.getCollection(collection).insertOne(document);
    return { ...document, _id: result.insertedId };
  }

  async insertMany(collection: string, documents: any[]): Promise<any[]> {
    const result = await this.getCollection(collection).insertMany(documents);
    return documents.map((doc, index) => ({
      ...doc,
      _id: result.insertedIds[index],
    }));
  }

  async findOne(collection: string, filter: any): Promise<any | null> {
    return await this.getCollection(collection).findOne(filter);
  }

  async findMany(
    collection: string,
    filter: any = {},
    options: any = {}
  ): Promise<any[]> {
    const cursor = this.getCollection(collection).find(filter);
    
    if (options.sort) {
      cursor.sort(options.sort);
    }
    if (options.limit) {
      cursor.limit(options.limit);
    }
    if (options.skip) {
      cursor.skip(options.skip);
    }

    return await cursor.toArray();
  }

  async updateOne(collection: string, filter: any, update: any): Promise<any> {
    const result = await this.getCollection(collection).updateOne(filter, {
      $set: update,
    });
    return result;
  }

  async updateMany(
    collection: string,
    filter: any,
    update: any
  ): Promise<number> {
    const result = await this.getCollection(collection).updateMany(filter, {
      $set: update,
    });
    return result.modifiedCount;
  }

  async deleteOne(collection: string, filter: any): Promise<boolean> {
    const result = await this.getCollection(collection).deleteOne(filter);
    return result.deletedCount > 0;
  }

  async deleteMany(collection: string, filter: any): Promise<number> {
    const result = await this.getCollection(collection).deleteMany(filter);
    return result.deletedCount;
  }

  async count(collection: string, filter: any = {}): Promise<number> {
    return await this.getCollection(collection).countDocuments(filter);
  }
}

